### PR TITLE
core/service: fix assertion when Type=dbus but BusName= is not specified

### LINKED ISSUE #19920: systemd (v247) crashes when Type=dbus unit is missing a BusName= attribute
**systemd version the issue has been seen with**
 > v247

**Used distribution**
 > NixOS

**Linux kernel version used** (`uname -a`)

 > `5.10.43`

**CPU architecture issue was seen on**
 > x86_64

**Expected behaviour you didn't see**
 The unit should be ignored & the user warned as it has an invalid definition.

**Unexpected behaviour you saw**
 systemd crashes during boot when the unit is pulled into multi-user.target

**Steps to reproduce the problem**
 1. Create a new unit of `Type=dbus`, make sure to not include any `BusName=` attribute
 2. Pull the unit into the multi-user.target
 3. Reboot
 4. See systemd crash once it loads the unit.

**Additional program output to the terminal or log subsystem illustrating the issue**

Here is what it looked like on one of my systems: 
image

Here is what it looks like when running a (NixOS) VM with a unit as defined above:

```text
machine # <<< NixOS Stage 2 >>>
machine #
machine # [    1.143637] EXT4-fs (vda): re-mounted. Opts: (null)
machine # [    1.144480] booting system configuration /nix/store/9i6gaiyl8fm9rsylabwbc3rpafzbs4jq-nixos-system-machine-21.11pre-git
machine # running activation script...
machine # setting up /etc...
machine # starting systemd...
machine # [    2.088761] systemd[1]: Inserted module 'autofs4'
machine # [    2.108056] systemd[1]: systemd 247 running in system mode. (+PAM +AUDIT -SELINUX +IMA +APPARMOR +SMACK -SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT -GNUTLS +ACL +XZ +LZ4 -ZSTD +SECCOMP +BLKID -ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified)
machine # [    2.110166] systemd[1]: Detected virtualization kvm.
machine # [    2.110659] systemd[1]: Detected architecture x86-64.
machine # [    2.111674] systemd[1]: Set hostname to <machine>.
machine # [    2.113483] systemd[1]: Initializing machine ID from random generator.
machine # [    2.158556] systemd-fstab-generator[444]: Checking was requested for "store", but it is not a device.
machine # [    2.162156] systemd-fstab-generator[444]: Checking was requested for "shared", but it is not a device.
machine # [    2.163653] systemd-fstab-generator[444]: Checking was requested for "xchg", but it is not a device.
machine # [    2.389087] systemd[1]: Assertion 'name' failed at src/core/unit.c:3496, function unit_watch_bus_name(). Aborting.
machine # [    2.390202] Core dump to |/bin/false pipe failed
machine # [    2.390716] systemd[1]: Caught <ABRT>, core dump failed (child 452, code=killed, status=6/ABRT).
machine # [    2.391563] systemd[1]: Freezing execution.
```

Unit file equivalent that I used:
```
[Unit]

[Service]

ExecStart=/some/shell/script/that/just/sleeps.sh
Type=dbus

[Install]
WantedBy=multi-user.target
```

### PR DESCRIPTION
**Pull request by yuwata**
Fixes #19920.

### CODE REVIEW

**Review by andir — APPROVED** [review]
Applied this patch onto my v247 tree and the bug described in #19920 disappeared.

During machine startup I now the following lines stating the situation as expected:
> machine # [    2.373669] systemd[1]: my.service: Service is of type D-Bus but no D-Bus service name has been specified. Refusing.
> machine # [    2.385169] systemd[1]: my.service: Cannot add dependency job, ignoring: Unit my.service has a bad unit file setting.

**Comment by yuwata** [comment]
@andir Thank you for testing the patch so quickly!

**Review by bluca — APPROVED** [review]
LGTM

**Comment by mbiebl** [comment]
Would it be good to add a test case for this?

**Comment by yuwata** [comment]
The previous commit was broken. Replaced completely new one. Also added a testcase. PTAL.

**Comment by mbiebl** [comment]
@evverx 
Do we have tests for feeding junk data to systemd as unit files? Overly large files, badly formatted, known broken ones, randomly generated junk,  etc.

File: test/units/testsuite-42.sh [hunk 1]
```diff
@@ -65,6 +65,10 @@ test -f /run/dbus1
 systemd-run --unit=dbus2.service --wait -p StandardOutput=tty -p StandardError=tty -p Type=dbus -p BusName=systemd.test.ExecStopPost -p ExecStopPost='/bin/touch /run/dbus2' true
 test -f /run/dbus2
 
+# https://github.com/systemd/systemd/issues/19920
+systemd-run --unit=dbus3.service --wait -p StandardOutput=tty -p StandardError=tty -p Type=dbus -p ExecStopPost='/bin/touch /run/dbus3' true \
+     && { echo 'unexpected success'; exit 1; }
+
 cat >/tmp/notify1.sh <<EOF
```
Comment (mbiebl) [inline test/units/testsuite-42.sh:72]: Does this properly catch the fact that PID 1 was crashing?

Comment (yuwata) [reply test/units/testsuite-42.sh:72]: If PID1 crashes, then the test freezes and hopefully CIs finish with timeout.

Comment (yuwata) [reply test/units/testsuite-42.sh:72]: And the journal should show some thing like the following:
```
Jun 15 03:18:01 H systemd[1]: Assertion 'name' failed at src/core/unit.c:3459, function unit_watch_bus_name(). Aborting.
Jun 15 03:18:01 H audit[530]: ANOM_ABEND auid=4294967295 uid=0 gid=0 ses=4294967295 pid=530 comm="systemd" exe="/usr/lib/systemd/systemd" sig=6 res=1
Jun 15 03:18:01 H systemd-coredump[531]: Due to PID 1 having crashed coredump collection will now be turned off.
Jun 15 03:18:01 H systemd-coredump[531]: 8;;man:core(5)^G[🡕]8;;^G Process 530 (systemd) of user 0 dumped core.
(snip)
```

**Comment by bluca** [comment]
> @evverx
> Do we have tests for feeding junk data to systemd as unit files? Overly large files, badly formatted, known broken ones, randomly generated junk, etc.

fuzzing should take care of that, all directives are fed fuzz data

**Comment by bluca** [comment]
> @evverx
> Do we have tests for feeding junk data to systemd as unit files? Overly large files, badly formatted, known broken ones, randomly generated junk, etc.

fuzzing should take care of that, all directives are fed fuzz data

**Comment by evverx** [comment]
> Do we have tests for feeding junk data to systemd as unit files? Overly large files, badly formatted, known broken ones, randomly generated junk, etc.

@mbiebl as far as I know there is one fuzz target covering `config_parse`. I'm not sure it triggers the code path mentioned in https://github.com/systemd/systemd/issues/19920. According to https://storage.googleapis.com/oss-fuzz-coverage/systemd/reports/20210613/linux/src/systemd/src/core/unit.c.html#L3491 `unit_add_node_dependency` has never been run on OSS-Fuzz

**Comment by boucman** [comment]
is this something to add to systemd-analyze verify ?

**Comment by bluca** [comment]
Doesn't look related, but looks new in bionic-arm64:

```
Assertion 'hashmap_put(s, UINT32_TO_PTR(__NR_ppoll + 1), INT_TO_PTR(EILSEQ)) >= 0' failed at src/test/test-seccomp.c:859, function test_load_syscall_filter_set_raw(). Aborting.
syscallrawseccomp terminated by signal ABRT.
Assertion 'wait_for_terminate_and_check("syscallrawseccomp", pid, WAIT_LOG) == EXIT_SUCCESS' failed at src/test/test-seccomp.c:873, function test_load_syscall_filter_set_raw(). Aborting.
```

**Comment by yuwata** [comment]
Oh, should be caused by 0643eb47a0418dc90d33853089dc9bc6ad63b0ca... cc @kraj.

**Comment by yuwata** [comment]
This should fix the issue https://github.com/systemd/systemd/pull/19935.

**Comment by bluca** [comment]
> > Do we have tests for feeding junk data to systemd as unit files? Overly large files, badly formatted, known broken ones, randomly generated junk, etc.
> 
> @mbiebl as far as I know there is one fuzz target covering `config_parse`. I'm not sure it triggers the code path mentioned in #19920. According to https://storage.googleapis.com/oss-fuzz-coverage/systemd/reports/20210613/linux/src/systemd/src/core/unit.c.html#L3491 `unit_add_node_dependency` has never been run on OSS-Fuzz

Interesting, maybe because it requires a valid type rather than just garbage to get there?

**Comment by evverx** [comment]
> Interesting, maybe because it requires a valid type rather than just garbage to get there?

@bluca my guess would be that apart from just parsing unit files the fuzzer should spawn a "fake" manager and do something useful to cover that part of the codebase.

**Pull request merged**